apollo-federation/src/operation/mod.rs

```diff
@@ -3922,6 +3648,400 @@ impl NamedFragments {
+    }
+
+    fn has_defer(&self) -> bool {
+        self.selections.values().any(|s| s.has_defer())
+    }
+
+    fn normalize_defer(self, normalizer: &mut DeferNormalizer) -> Result<Self, FederationError> {
+        let Self {
+            schema,
+            type_position,
+            selections,
+        } = self;
+        Arc::unwrap_or_clone(selections)
+            .into_iter()
+            .map(|(_, sel)| sel.normalize_defer(normalizer))
+            .try_collect()
+            .map(|selections| Self {
+                schema,
+                type_position,
+                selections: Arc::new(selections),
+            })
+    }
+}
+
+impl Operation {
+    fn has_defer(&self) -> bool {
+        self.selection_set.has_defer()
+            || self
+                .named_fragments
+                .fragments
+                .values()
+                .any(|f| f.has_defer())
+    }
+
+    /// Create a new operation without @defer directive applications.
+    pub(crate) fn without_defer(mut self) -> Result<Self, FederationError> {
+        if self.has_defer() {
+            let named_fragments = self.named_fragments.without_defer(DeferFilter::All)?;
+            self.selection_set = self
+                .selection_set
+                .without_defer(DeferFilter::All, &named_fragments)?;
+            self.named_fragments = named_fragments;
+        }
+        debug_assert!(!self.has_defer());
+        Ok(self)
+    }
+
+    /// Create a new operation without specific @defer(label:) directive applications.
+    pub(crate) fn reduce_defer(
+        mut self,
+        labels: &IndexSet<String>,
+    ) -> Result<Self, FederationError> {
+        if self.has_defer() {
+            let named_fragments = self
+                .named_fragments
+                .without_defer(DeferFilter::Labels(labels))?;
+            self.selection_set = self
+                .selection_set
+                .without_defer(DeferFilter::Labels(labels), &named_fragments)?;
+            self.named_fragments = named_fragments;
+        }
+        Ok(self)
+    }
+
+    /// Returns this operation but modified to "normalize" all the @defer applications.
+    ///
+    /// "Normalized" in this context means that all the `@defer` application in the resulting
+    /// operation will:
+    ///  - have a (unique) label. Which implies that this method generates a label for any `@defer`
+    ///    not having a label.
+    ///  - have a non-trivial `if` condition, if any. By non-trivial, we mean that the condition
+    ///    will be a variable and not an hard-coded `true` or `false`. To do this, this method will
+    ///    remove the condition of any `@defer` that has `if: true`, and will completely remove any
+    ///    `@defer` application that has `if: false`.
+    pub(crate) fn with_normalized_defer(mut self) -> Result<NormalizedDefer, FederationError> {
+        if self.has_defer() {
+            // TODO(@goto-bus-stop): does this need named fragment handling?
```

Comment:
Should solve this and add tests before merge

Reply:
Okay, I'll explicitly not handle this, as it requires quite a bit of work just for potentially helping `.reuse_fragments()`. It should be fine since we don't track which named fragment an expanded fragment comes from